Доработана функциональность по техническому заданию 6 спринта

asdbl opened the pull request.

asdbl committed:
Добавить возможность хранить более 10 задач в истории.
Добавить отображение только последнего просмотра задачи.
Изменить модификаторы доступа сеттеров.

asdbl committed:
Исправить ошибку теста CheckStyle на GitHub

asdbl committed:
Исправить ошибку теста CheckStyle на GitHub

asdbl committed:
Исправить ошибку теста CheckStyle на GitHub

asdbl committed:
Исправить ошибку теста CheckStyle на GitHub

TyutterinYakov commented on src/manager/Managers.java:
(предложение) Так как класс не имеет состояния(все static), значит создание объектов этого класса не имеет смысла. Можно объявить в нем приватный конструктор без параметров, чтобы избежать создания бессмысленных объектов этого класса. Это тонкость, которую хорошо было бы понять) То есть - если у класса нет публичного конструктора - создание объекта этого класса невозможно(ну почти, в java есть некоторый механизм, который может помочь это обойти, но об этом позже)) Когда ты создаешь класс, в нем по дефолту есть публичный конструктор без параметров, который в коде не видно, но он есть. Если мы создаем приватный, мы его перекрываем, а значит запрещаем создавать объекты этого класса

TyutterinYakov commented on src/manager/historyManager/HistoryNode.java:
```diff
@@ -0,0 +1,15 @@
+
+import task.Task;
+
+public class HistoryNode<T> {
```
Для соблюдения инкапсуляции - поля класса должны быть приватными, доступ к полям класса должен осуществляться через геттеры и сеттеры)

(предложение) Либо же можно было бы сделать его внутреннего приватного класса, внутри менеджера истории, так как он используется лишь для его внутренней реализации
https://docs.oracle.com/javase/tutorial/java/javaOO/nested.html

TyutterinYakov commented on src/manager/historyManager/InMemoryHistoryManager.java:
```diff
@@ -2,32 +2,74 @@
-import java.util.List;
+import java.util.*;
 
 public class InMemoryHistoryManager implements HistoryManager {
```
Для соблюдения инкапсуляции - поля класса должны быть приватными)

TyutterinYakov commented on src/manager/historyManager/InMemoryHistoryManager.java:
```diff
@@ -2,32 +2,74 @@
-    private int historyListIndex = 0;
+    private final Map<Integer, HistoryNode<Task>> historyMap;
+    HistoryNode<Task> lastNode;
+    HistoryNode<Task> currentNode;
```
Вынос этой переменной на уровень полей класса - не совсем верно, так как она требуется лишь внутри метода

TyutterinYakov commented on src/manager/historyManager/InMemoryHistoryManager.java:
```diff
@@ -2,32 +2,74 @@
+
+    @Override
+    public void remove(int id) {
+        removeNode(historyMap.get(id));
```
Можно было бы сразу же использовать `removeNode(historyMap.remove(id))`, таким образом мы не будем лишний раз обращаться к мапе

TyutterinYakov commented on src/manager/taskManager/InMemoryTaskManager.java:
```diff
@@ -2,17 +2,14 @@
-public class InMemoryTaskManager implements TaskManager {
-    private int id = 1;
+public class InMemoryTaskManager extends StatusSetter implements TaskManager {
+    private static int id = 1;
```
Поле не должно быть `static`, так как каждый объект менеджера задач должен иметь свое значение последнего идентификатора

TyutterinYakov commented on src/manager/taskManager/InMemoryTaskManager.java:
```diff
@@ -2,17 +2,14 @@
-    private int id = 1;
+public class InMemoryTaskManager extends StatusSetter implements TaskManager {
+    private static int id = 1;
     private final HashMap<Integer, Task> tasks;
```
Раз ты познакомился с интерфейсами и полиморфизмом, может быть стоит теперь использовать интерфейс `Map`, чтобы не привязывать к конкретной его реализации, а работать лишь с предоставляемым API) Все места не отмечал, но думаю, что ты разберешься)

TyutterinYakov commented on src/manager/taskManager/TaskManager.java:
```diff
@@ -8,18 +8,31 @@
+
     void update(Subtask subtask);
+
     ArrayList<Task> getTasks();
```
Раз ты познакомился с интерфейсами и полиморфизмом, может быть стоит теперь использовать интерфейс `List`, чтобы не привязывать к конкретной его реализации, а работать лишь с предоставляемым API) Все места не отмечал, но думаю, что ты разберешься)

TyutterinYakov commented on src/manager/taskManager/InMemoryTaskManager.java:
Присутствует удаление из истории не во всех местах, где удаляются задачи/подзадачи/эпики

TyutterinYakov commented on src/task/StatusSetter.java:
```diff
@@ -0,0 +1,7 @@
+package task;
+
+public class StatusSetter {
```
Действительно ли этот класс требуется?) Можно было бы вызывать сразу сеттер из соответствующего объекта класса

TyutterinYakov commented on src/task/Task.java:
```diff
@@ -9,6 +11,7 @@ public class Task {
     private Status status;
 
     public Task(String taskName, String taskDescription, Status status) {
+        setId(InMemoryTaskManager.getId());
```
Установка идентификатора должна происходить в методах создания задачи/подзадачи/эпика

TyutterinYakov commented on src/manager/historyManager/InMemoryHistoryManager.java:
```diff
@@ -2,32 +2,74 @@
+    }
+
+   private ArrayList<Task> getTask() {
+        ArrayList<Task> taskList = new ArrayList<>();
```
Можно было бы использовать `LinkedList`, так как в нем операции вставки, с чем есть проблема у `ArrayList`, выполняются за O(1). Можно выделить сразу требуемый размер(new `ArrayList<>(historyMap.size()))` - это решит проблему вставки)
Будет очень полезно почитать про то, как работают методы этих классов, это спрашивают на собеседованиях и действительно пригождается в работе. https://www.javatpoint.com/difference-between-arraylist-and-linkedlist

TyutterinYakov commented on src/manager/historyManager/InMemoryHistoryManager.java:
```diff
@@ -2,32 +2,74 @@
         }
-        if (historyList.size() == INITIAL_CAPACITY) {
-            historyList.removeFirst();
+        if (historyMap.containsKey(task.getId())) {
```
Можно было бы реализовать эту логику более оптимально, лишь с одним обращением к мапе)
`removeNode(historyMap.remove(task.getId()));`
А саму проверку на то, что пришел не `null` занести в метод `removeNode`, чтобы не производить ее в других местах

TyutterinYakov commented on src/manager/historyManager/InMemoryHistoryManager.java:
```diff
@@ -2,32 +2,74 @@
+    }
+
+   private void linkLast(Task task) {
+        HistoryNode<Task> l = lastNode;
```
Можно было бы не вводить дополнительную переменную и не дублировать код) Например:
```
HistoryNode<Task> newNode = new HistoryNode<>(l, task, null);
if (firstNode == null) {
    firstNode = newNode;
} else {
    lastNode.next = newNode;
}
lastNode = newNode;
```

TyutterinYakov commented on src/manager/historyManager/InMemoryHistoryManager.java:
```diff
@@ -2,32 +2,74 @@
     }
+
+   private void removeNode(HistoryNode<Task> node) {
+        if (node.prevNode != null) {
```
Хорошо было бы проверить значение `node` на `null`, чтобы исключить проблему с NPE)

Также хорошо было бы ввести константы для значений `node.prev` и `node.next`, чтобы не дублировать код и упростить его восприятие)

TyutterinYakov reviewed (commented):
Добрый день! В целом - хорошо) Отметил то, что следует исправить и то, что можно сделать лучше) Хорошего дня! Если возникнут вопросы по ревью - пиши в пачку, буду рад помочь🌅

asdbl committed:
Сделать класс HistoryNode внутренним классом InMemoryHistoryManager.
Удалить класс StatusSetter.
Сделать сеттеры public.
Добавить приватный конструктор без параметров в Managers.
Сделать поля InMemoryHistoryManager приватными.
Добавить удаление истории при вызове метода clear в InMemoryTaskManager.
Заменить ArrayList на LinkedList в методе getTask() в классе InMemoryHistoryManager.
Убрать дублирование кода в методе linkLast(Task task) в классе InMemoryHistoryManager.
Добавить проверку на null в методе removeNode в классе InMemoryHistoryManager.

asdbl committed:
Исправить ошибку CheckStyle на GitHub.

TyutterinYakov commented on src/manager/historyManager/InMemoryHistoryManager.java:
```diff
@@ -2,32 +2,95 @@
-        }
-        if (historyList.size() == INITIAL_CAPACITY) {
-            historyList.removeFirst();
+        if (historyMap.containsKey(task.getId())) {
```
Проверка `if (historyMap.containsKey(task.getId())) {` не имеет смысла, так как теперь в методе `removeNode` есть проверка на `null`

TyutterinYakov commented on src/manager/historyManager/InMemoryHistoryManager.java:
```diff
@@ -2,32 +2,95 @@
 
     @Override
     public void add(Task task) {
-        if (task == null) {
```
Эту проверку следует вернуть, чтобы избежать проблем с NPE, если будет запрошена несуществующая задача в менеджере задач

TyutterinYakov commented on src/manager/taskManager/InMemoryTaskManager.java:
```diff
@@ -2,23 +2,20 @@
     private final HistoryManager historyManager = Managers.getDefaultHistory();
 
-    public InMemoryTaskManager(HashMap<Integer, Task> tasks, HashMap<Integer, Epic> epics, HashMap<Integer, Subtask> subtasks) {
+    public InMemoryTaskManager(Map<Integer, Task> tasks, Map<Integer, Epic> epics, Map<Integer, Subtask> subtasks) {
```
Лучше произвести инициализацию мап в месте объявления, а не передавать их из вне, чтобы избежать их модификации вне менеджера задач

TyutterinYakov reviewed (approved):
Добрый вечер! Поздравляю с успешной сдачей спринта :) Сил и терпения на этом сложном, но интересном пути! Успехов в обучении и прекрасного вечера!🌃🎇

asdbl committed:
Удалить проверку historyMap.containsKey в методе add классе InMemoryHistoryManager.
Добавить проверку на null в метод add в классе InMemoryHistoryManager.
